src/TokenFactory.sol

```diff
@@ -0,0 +1,123 @@
+import {PoolAddress} from "./libraries/PoolAddress.sol";
+import {Initializable} from "./base/Initializable.sol";
+import {ERC20Standard} from "./base/ERC20Standard.sol";
+import {Ownable} from '@openzeppelin/contracts/access/Ownable.sol';
+import '@openzeppelin/contracts/proxy/Clones.sol';
+
+
+contract TokenFactory is Ownable, ITokenFactory, Initializable {
+    uint256 private nonce = 1;
+    address public override erc20WhitelistImplementation;
+    address public override oracleWhitelistImplementation;
+    address public override uniswapV3Factory;
+
+    constructor() {
+        transferOwnership(tx.origin);
+    }
+
+    function initialize(address _erc20WhitelistImplementation, address _oracleWhitelistImplementation, address _uniswapV3Factory) external whenNotInitialized() {
+        erc20WhitelistImplementation = _erc20WhitelistImplementation;
+        oracleWhitelistImplementation = _oracleWhitelistImplementation;
+        uniswapV3Factory = _uniswapV3Factory;
+    }
+
+    // Set the implementation address for the ERC20Whitelist contract
+    function setERC20WhitelistImplementation(address _erc20WhitelistImplementation) external override onlyOwner {
+        emit ERC20WhitelistImplementationSet(erc20WhitelistImplementation, _erc20WhitelistImplementation);
+        erc20WhitelistImplementation = _erc20WhitelistImplementation;
+    }
+
+    // Set the implementation address for the OracleWhitelist contract
+    function setOracleWhitelistImplementation(address _oracleWhitelistImplementation) external override onlyOwner {
+        emit OracleWhitelistImplementationSet(oracleWhitelistImplementation, _oracleWhitelistImplementation);
+        oracleWhitelistImplementation = _oracleWhitelistImplementation;
+    }
+
+    function createERC20WhitelistToken(IERC20Whitelist.InitializeParams calldata params) external override returns (address token) {
+        // adding the salt to the address to make it cross-chain unique
+        bytes32 salt = keccak256(abi.encodePacked(
+                ChainId.get(),
+                nonce++
+            ));
+        token = Clones.cloneDeterministic(erc20WhitelistImplementation, salt);
+        IERC20Whitelist(token).initialize(params);
+        emit TokenCreated(token, params);
+    }
+
+    function createOracleWhitelist(IOracleWhitelist.InitializeParams calldata params) external override returns (address whitelistAddress) {
+        // adding the salt to the address to make it cross-chain unique
+        bytes32 salt = keccak256(abi.encodePacked(
+                ChainId.get(),
+                nonce++
+            ));
+        whitelistAddress = Clones.cloneDeterministic(oracleWhitelistImplementation, salt);
+        IOracleWhitelist(whitelistAddress).initialize(params);
+        emit OracleWhitelistCreated(whitelistAddress, params);
+    }
+
+    function createWhitelistContracts(CreateWhitelistContractsParams calldata params) external override returns (address token, address whitelistAddress) {
```

Comment:
`createWhitelistContracts` seems named not really clearly 😂

Reply:
`createTokenAndWhitelist`?

Reply:
comments could help too